### PR TITLE
Adapt tests to reuse Orchestrator

### PR DESCRIPTION
**Pull request by ashao**
Tests which needed to launch an Orchestrator were spinning up and shutting down their own instances. This led to a number of cases where a single test failing would cascade into failures of other tests. Additionally, this also meant that a significant amount of time in the tests was spent waiting for Orchestrators to launch.

This PR adds a session-scoped fixture that returns an Orchestrator. Most tests which use an Orchestrator have been updated to use this fixture; the remaining for various reasons still need to spin up their own (for example the multiple database tests need to have a named Orchestrator).

This PR is not 100% complete with the following to be done
- Add another fixture that flushes the database after every test
- Undergo style checks
- Fix failing tests

### CODE REVIEW

File: conftest.py
Comment (ashao) [inline conftest.py:232]: I think this must have come in as a weird merge commit. Will delete

File: conftest.py
Comment (ankona) [inline conftest.py:269]: Depending on when this gets merged, you might be able to swap to using [`smartsim._core.utils.network.find_free_port`](https://github.com/CrayLabs/SmartSim/blob/dragon_launcher/smartsim/_core/utils/network.py#L107-L125)

File: conftest.py
Comment (ankona) [inline conftest.py:784]: 👍 for use of scope.

File: smartsim/_core/config/config.py
Comment (ankona) [inline smartsim/_core/config/config.py:184]: 1. I'm really disliking the mixing of test and production configuration in `config.py`. It might be better to set test values in `conftest.py` - or in this case, right in the fixture. 

2. Why do we limit ourselves here to 10 available ports? It could be 100 or 1000 if you're trying to avoid an infinite loop. If we actually need 10 databases and there is a port or two unavailable, the tests just start failing.

File: smartsim/database/orchestrator.py
Comment (MattToast) [inline smartsim/database/orchestrator.py:361]: Is the try catch needed here? It looks like the hosts attr (for some reason) can throw this `SSDBFilesNotParseable` error

**Review by MattToast — COMMENTED** [review]
Two small questions about intent that I wanted to write down so that I could cover them later while I pick up this PR

File: smartsim/_core/control/controller.py
Comment (MattToast) [inline smartsim/_core/control/controller.py:779]: Why was this sleep removed?

File: tests/test_experiment.py
Comment (MattToast) [inline tests/test_experiment.py:82]: Why was this test removed?

File: smartsim/experiment.py
Comment (MattToast) [inline smartsim/experiment.py:154]: Similar question here: why did this PR touch this code path?

File: conftest.py
Comment (MattToast) [inline conftest.py:254]: I'm a little confused on the logic here: It looks like the idea is that this socket is going to try to connect to this addr/port, and if there is some error we assume that it is because there is no socket bound there to accept the connection, and therefore the port must be free.

However, errors could come from a variety of problems (e.g. there is a socket bound but it is not listening, there is a socket bound and it is listening but it ran out of buffer space for pending new unaccepted connections, etc.) where the port may not actually be available.

Should this be trying to bind instead??

Comment (MattToast) [reply conftest.py:254]: Switched to `bind` to avoid port conflicts

File: smartsim/database/orchestrator.py
Comment (MattToast) [reply smartsim/database/orchestrator.py]: Leftover for a previous refactor attempt that was reverted. The try catch block was correctly moved to wrap `self.hosts`

File: tests/test_experiment.py
Comment (MattToast) [reply tests/test_experiment.py:82]: This is something we want to support, but was deemed out of scope for this PR. Reverting this change

File: smartsim/experiment.py
Comment (MattToast) [reply smartsim/experiment.py:154]: Re-adding this test per reverted change

File: conftest.py
Comment (MattToast) [reply conftest.py]: Current plan is this PR will likely go in before the dragon work, so this fn is not yet defined, but we should consider using it as soon as the dragon work is merged!

File: smartsim/_core/config/config.py
Comment (MattToast) [reply smartsim/_core/config/config.py]: 1. Totally agree, we are already pushing the number of public methods on the `Config` class! I'll write up a ticket to make multiple configuration class for different areas of configuration!!
1. After talking with Andrew, it seems like you are right! The 10 port limit was deliberately chosen to prevent an infinite loop, and 10 was chosen slightly arbitrarily just to keep the iteration to an order of 10s. Sounds like there is some wiggle room to be debated, but I actually kinda like that it doesn't check an infinite number of ports as if orcs start leaking from tests we have a way to back track to the offending test. Willing to be debated on this point though!!

**Review by AlyssaCote — CHANGES_REQUESTED** [review]
This is awesome! And I really like how it was implemented. I have a couple questions and there's a test file that needs some love, but that's it!

File: smartsim/_core/config/config.py
Comment (ankona) [reply smartsim/_core/config/config.py]: I'm ok w/any number that feels less likely to randomly break than 10 - anything between 25 and 1000 is still arbitrary but also feels safer!

**Review by ankona — APPROVED** [review]
This PR is like a dream come true. I can't wait for my speedy test results.

File: conftest.py
Comment (ankona) [reply conftest.py:414]: @AlyssaCote The fixtures [have scopes applied](https://github.com/CrayLabs/SmartSim/pull/567/files#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59R246) that will ensure they only get called once per run. That lets pytest return the yielded result to every test until the scope expires!

Comment (AlyssaCote) [reply conftest.py:414]: @ankona Thanks for the explanation!! Makes sense!

File: smartsim/_core/config/config.py
Comment (MattToast) [reply smartsim/_core/config/config.py]: That makes sense, and I like the reasoning for a higher number! I'll bump it to 25, and we can increase it in future as we see fit!

File: conftest.py
Comment (MattToast) [reply conftest.py:568]: We probably should, this was 100% me being lazy and keeping the existing impl. Will change the type of this param to `int | None` where `None` is a sentinel value to find a free port!

File: tests/test_fixtures.py
Comment (MattToast) [reply tests/test_fixtures.py:42]: Uhhh... good question! I've never written testing code for testing code before. I'll defer to @ashao on this one for intention.

File: smartsim/database/orchestrator.py
Comment (amandarichardsonn) [inline smartsim/database/orchestrator.py:404]: No need to add the rtype here I believe!

File: smartsim/database/orchestrator.py
Comment (amandarichardsonn) [inline smartsim/database/orchestrator.py:400]: I would add this property in doc/api/smartsim_api around line 389 unless this function shouldnt be user facing, then I would add a _ infront or else it will show up on readthedocs

Comment (amandarichardsonn) [reply smartsim/database/orchestrator.py:400]: do you mind adding the batch property as well?

File: smartsim/error/errors.py
Comment (amandarichardsonn) [inline smartsim/error/errors.py:90]: Thanks for adding a description! I am lost without them fr

File: conftest.py
Comment (amandarichardsonn) [inline conftest.py:263]: Is a ticket created for this? If not, possibly might forget!

File: doc/changelog.md
Comment (amandarichardsonn) [inline doc/changelog.md:105]: I notice two updates to the Detailed Notes section but one update to the Description!

File: smartsim/_core/control/controller.py
Comment (amandarichardsonn) [inline smartsim/_core/control/controller.py:790]: ticket created 👀

**Review by amandarichardsonn — APPROVED** [review]
LGTM, will take a second pass in the morning! Just some super small comments

File: smartsim/database/orchestrator.py
Comment (MattToast) [reply smartsim/database/orchestrator.py]: `:rtype:` is gone!

File: doc/changelog.md
Comment (MattToast) [reply doc/changelog.md:105]: Yeah the change to "Description" and one of the "Detailed Notes" edits were just minor typo/grammar things.

The actual "Detailed Note" I added I figured fell under the category of "Minor enhancements to test suite" which is already listed in the "Description" section, so I just added something the details. LMK if you think this warrants its on entry in the "Description" and I can add something there!!

File: smartsim/database/orchestrator.py
Comment (MattToast) [reply smartsim/database/orchestrator.py:400]: Added!

File: smartsim/_core/control/controller.py
Comment (MattToast) [reply smartsim/_core/control/controller.py:790]: [Added](https://github.com/CrayLabs/Poseidon/issues/690)!!

File: conftest.py
Comment (MattToast) [reply conftest.py:263]: [Added](https://github.com/CrayLabs/Poseidon/issues/691)!!

File: tests/test_fixtures.py
Comment (ashao) [reply tests/test_fixtures.py:42]: How meta do we want to go? But yes the intention of this was to make sure that fixture works as expected. I'm imagining us bashing our head against something related to the fixture and just being able to do:
```
pytest test_fixtures.py::test_single_db_fixture
```
and save us some time down the line. Happy to be argued down, but in general I think it's not crazy to have a some sanity checking of testing code since it helps to debug tests

Comment (AlyssaCote) [reply tests/test_fixtures.py:42]: I definitely don't hate the idea of being able to sanity check that the fixtures are working as expected.

**Comment by ashao** [comment]
Note that Al noted a defect which causes cascading failures if the database was stopped on purpose or because of a database error. A fix is coming in

File: conftest.py
Comment (AlyssaCote) [inline conftest.py:874]: Does the `database_registry` hold the name of the experiment instead of the database?

**Review by AlyssaCote — COMMENTED** [review]
This looks really great! I have a few general questions rather than specific fix requests, but once those are cleared up this should be good to go!

File: conftest.py
Comment (ashao) [reply conftest.py:789]: Ahhh nice catch! Thanks for that

File: tests/test_collectors.py
Comment (ashao) [reply tests/test_collectors.py:45]: Mainly because @ankona typehinted everything else in this file. We should probably do the same for the other tests at some point

File: conftest.py
Comment (ashao) [reply conftest.py:874]: It's supposed to be the identifier for a specific "type" of orchestrator that a test requests. Since we can't launch a database outside of an experiment, the two are synonymous right now.

Comment (AlyssaCote) [reply conftest.py:874]: Makes sense!

File: conftest.py
Comment (AlyssaCote) [inline conftest.py:780]: Totally missed this on the first pass through, but can we add type annotations to the new functions throughout this file?

**Review by AlyssaCote — APPROVED** [review]
LGTM!!

**Pull request merged**
